### PR TITLE
Add attack level range to combat level plugin

### PR DESCRIPTION
**Pull request by Nightfirecat**
This plugin adds a PVP-world-like combat level range text to the
wilderness level, displaying the range of combats you can attack at any
given wilderness level.

Attack level range plugin demonstration

Closes runelite/runelite#3103
Closes runelite/runelite#3318

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/attacklevelrange/AttackLevelRangePlugin.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+		}
+		if (originalSkullContainerPosition == -1)
+		{
+			originalSkullContainerPosition = skullContainer.getRelativeY();
+		}
+
+		final int wildernessLevel = Integer.parseInt(m.group(1));
+		final int combatLevel = client.getLocalPlayer().getCombatLevel();
+
+		wildernessLevelWidget.setText(wildernessLevelText + "<br>" + (combatLevel - wildernessLevel) + "-" + (combatLevel + wildernessLevel));
```
Comment (bmiddle) [inline runelite-client/src/main/java/net/runelite/client/plugins/attacklevelrange/AttackLevelRangePlugin.java]: You might want to check if the upper range is greater than 126 here.

Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/attacklevelrange/AttackLevelRangePlugin.java]: Good point. Fixed

**Comment by deathbeam** [comment]
Imo this could be in Combat Level plugin

**Comment by Nightfirecat** [comment]
@deathbeam I've moved this to the combat level plugin behind a config.

File: runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelPlugin.java [hunk 2]
```diff
@@ -25,33 +25,72 @@
...
 @PluginDescriptor(
 	name = "Combat Level",
-	description = "Show a more accurate combat level in Combat Options panel"
+	description = "Show a more accurate combat level in Combat Options panel and other combat level functions",
+	tags = {"wilderness", "attack", "range"}
 )
 public class CombatLevelPlugin extends Plugin
 {
-	private final DecimalFormat decimalFormat = new DecimalFormat("#.###");
+	private static final DecimalFormat decimalFormat = new DecimalFormat("#.###");
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelPlugin.java]: UPPER_CASE

Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelPlugin.java]: Done, and changed my inspections settings so you shouldn't have to bother me about this again. :laughing:

File: runelite-api/src/main/java/net/runelite/api/events/WidgetTextChanged.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.api.events;
+
+import lombok.Data;
+import net.runelite.api.widgets.Widget;
+
+@Data
+public class WidgetTextChanged
```
[the remark this answers is not in the document]
Comment (Nightfirecat) [reply runelite-api/src/main/java/net/runelite/api/events/WidgetTextChanged.java]: Done

**Comment by deathbeam** [comment]
+ travis failing

**Comment by Nightfirecat** [comment]
Thanks, fixed that--had to use a constructor for the event instead of `setWidget` and `setText`.

**Comment by Nightfirecat** [comment]
This PR now closes #3439, as I realized the new WidgetTextChanged event would fulfill that request.

**Comment by Nightfirecat** [comment]
Upon discussion in Discord, I'll be rewriting this later today to remove the `WidgetTextChanged` event and just do a `widget.setText` call in a GameTick subscriber.

**Comment by Nightfirecat** [comment]
Never mind on the GameTick change; the script which updates the wilderness level widget is called much more often, rendering anything happening only once per game tick useless.

**Comment by Nightfirecat** [comment]
I've rewritten this PR to add the PVP widget building script and to add a callback before its final `return`, and to run the text and position changes at that time. This eliminates the need to add the `WidgetTextChanged` event.

**Comment by Adam-** [comment]
This script needs to be updated for the script changes

**Comment by Nightfirecat** [comment]
Rebased with new script opts

**Pull request merged**